rpcs3/Emu/Cell/PPUThread.cpp

```diff
@@ -2013,17 +2037,278 @@ extern void ppu_initialize()
+
+			for (auto&& entry : fs::dir(dir_queue[i]))
+			{
+				if (Emu.IsStopped())
+				{
+					i = -1;
+					file_queue.clear();
+					break;
+				}
+
+				if (entry.is_directory)
+				{
+					if (entry.name != "." && entry.name != "..")
+					{
+						dir_queue.emplace_back(dir_queue[i] + entry.name + '/');
+					}
+
+					continue;
+				}
+
+				// Check .sprx, .prx, .self, .elf filenames
+				if (exec_type result = [&]()
+				{
+					std::string filename = fmt::to_upper(entry.name);
+
+					if (filename.ends_with(".SPRX"sv) || filename.ends_with(".PRX"sv))
+					{
+						return exec_type::prx;
+					}
+
+					if (filename.ends_with(".SELF"sv) || filename.ends_with(".ELF"sv))
+					{
+						return exec_type::ovl;
+					}
+
+					return exec_type::none;
+				}(); result != exec_type::none)
+				{
+					// Skip already loaded modules or HLEd ones
+					if (dir_queue[i] == firmware_sprx_path)
+					{
+						bool ignore = false;
+
+						for (auto* obj : prx_list)
+						{
+							if (obj->name == entry.name)
+							{
+								ignore = true;
+							}
+						}
+
+						if (ignore)
+						{
+							continue;
+						}
+
+						if (g_cfg.core.libraries_control.get_set().count(entry.name + ":lle"))
+						{
+							// Force LLE
+							ignore = false;
+						}
+						else if (g_cfg.core.libraries_control.get_set().count(entry.name + ":hle"))
+						{
+							// Force HLE
+							ignore = true;
+						}
+						else
+						{
+							extern const std::map<std::string_view, int> g_prx_list;
+
+							// Use list
+							ignore = g_prx_list.count(entry.name) && g_prx_list.at(entry.name) != 0;
+						}
+
+						if (ignore)
+						{
+							continue;
+						}
+					}
+
+					file_queue.emplace_back(dir_queue[i] + entry.name, result);
+				}
+			}
+		}
+
+		g_progr = "Compiling PPU modules";
+
+		atomic_t<usz> fnext = 0;
+
+		shared_mutex sprx_mtx;
+
+		named_thread_group workers("SPRX Worker ", std::min<u32>(Emu.GetMaxThreads(), ::size32(file_queue)), [&]
+		{
+			for (usz func_i = fnext++; func_i < (!Emu.IsStopped() ? file_queue.size() : 0); func_i = fnext++)
+			{
+				const auto& path = std::as_const(file_queue)[func_i].first;
+
+				sys_log.notice("Trying to load SPRX: %s", path);
+
+				// Load MSELF or SPRX
+				fs::file src{path};
+
+				// Some files may fail to decrypt due to the lack of klic
+				src = decrypt_self(std::move(src));
+
+				elf_error err{};
+
+				switch (file_queue[func_i].second)
+				{
+				case exec_type::prx:
+				{
+					const ppu_prx_object obj = src;
+
+					if (obj == elf_error::ok)
+					{
+						std::unique_lock lock(sprx_mtx);
+
+						if (auto prx = ppu_load_prx(obj, path))
+						{
+							lock.unlock();
+							ppu_initialize(*prx);
+							idm::remove<lv2_obj, lv2_prx>(idm::last_id());
+							lock.lock();
+							ppu_unload_prx(*prx);
+							lock.unlock();
+							ppu_finalize(*prx);
+							g_progr_fdone++;
+							continue;
+						}
+					}
+
+					err = obj;
+					break;
+				}
+				case exec_type::ovl:
```

Comment:
Why letting workers compile OVL if it's just like normal executable can only be compiled sequentially?

Reply:
I think MGS4 can load multiple at once, it depends on the addresses they are using.

Reply:
Also I may shuffle it differently so some threads compile OVL while other PRX.